design/01-proc-macro.md

```diff
@@ -0,0 +1,290 @@
+    - Emits compiler diagnostics.
+    - Can emit additional arbitrary data structure along generated file.
+
+Macro plugins are executed on each node in the Cairo AST.
+
+Custom plugins can be fed to the compiler via [`RootDatabaseBuilder::with_plugin_suite` API](https://github.com/starkware-libs/cairo/blob/8290898d602f87fb5bc3a18f28a42430cf6d3fdc/crates/cairo-lang-compiler/src/db.rs#L96). It takes a [`PluginSuite` object](https://github.com/starkware-libs/cairo/blob/8290898d602f87fb5bc3a18f28a42430cf6d3fdc/crates/cairo-lang-semantic/src/plugin.rs#L18-L25), which defines a set of objects implementing the [`MacroPlugin` trait](https://github.com/starkware-libs/cairo/blob/8290898d602f87fb5bc3a18f28a42430cf6d3fdc/crates/cairo-lang-defs/src/plugin.rs#L80).
+
+Scarb currently simply assumes a fixed, predefined plugin sets for `lib`, `starknet-contract` and `test` targets. This interface is not stable and is subject to changes in future compiler versions.
+
+### Plugins API in combination with Scarb
+
+Because of the use of shared reference to [`SyntaxGroup`], this API is hostile to attempts to externalize plugins into separate binaries/processes that communicate via some kind of RPC. The fact that AST structure is not stable does not help either. On the other hand, access to [`SyntaxGroup`] allows for powerful operations and is highly desired.
+
+The only feasible way of loading plugins is to load them directly into Scarb compiler's memory. This brings another problem though: Rust does not have a stable ABI, and thus such system would have to ensure that the plugin library was compiled with **exactly** the same Rust toolchain, for example `1.68.1-x86_64-pc-windows-msvc`.
+
+### Declaring package dependencies on plugins
+
+Scarb allows packages to declare dependency on a plugin through the same mechanism as on standard Cairo libraries. The plugin must be defined as a Scarb package defining `cairo-plugin` target in the package manifest.
+
+Since currently there is no mechanism to supply plugins implementation, it must be compiled into Scarb itself.
+The `starknet` and `test_plugin` plugins are supplied this way.
+
+## Guide-level explanation
+
+This document proposes an introduction of new compiler plugins type, called procedural macros, as described below.
+
+### Procedural macro user perspective
+
+To use a procedural macro, a Cairo programmer will have to:
+- Declare a dependency on a package, that implements the procedural macro, by adding it to the `dependencies` field in the Scarb manifest file.
+- Use the procedural macro in Cairo code, by calling it, or adding an attribute or derive macro to a Cairo item.
+
+The overall complexity of the actual procedural macro execution should be hidden from the user abstraction level.
+In particular, this two steps can be performed without any knowledge of Rust, or even the fact that the procedural macro is implemented in Rust.
```

Comment:
New macro downloads will require having Rust compiler installed though. This is worth noting here imho